modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java

```diff
@@ -0,0 +1,327 @@
+                            List.of(put(startBuildIndexKey, BYTE_EMPTY_ARRAY)),
+                            putPartitionBuildIndexOperations(indexId, partitions)
+                    ),
+                    List.of(noop())
+            );
+        });
+    }
+
+    private CompletableFuture<?> onIndexDrop(DropIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            int indexId = parameters.indexId();
+
+            int partitions = getPartitionCountFromCatalog(indexId, parameters.catalogVersion() - 1);
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
+
+            return metaStorageManager.invoke(
+                    exists(startBuildIndexKey),
+                    concat(
+                            List.of(remove(startBuildIndexKey)),
+                            removePartitionBuildIndexOperations(indexId, partitions)
+                    ),
+                    List.of(noop())
+            );
+        });
+    }
+
+    private CompletableFuture<?> onIndexAvailable(MakeIndexAvailableEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            ByteArray startBuildIndexKey = startBuildIndexKey(parameters.indexId());
+
+            return metaStorageManager.invoke(exists(startBuildIndexKey), remove(startBuildIndexKey), noop());
+        });
+    }
+
+    private CompletableFuture<?> onUpdatePartitionBuildIndexKey(WatchEvent event) {
+        return inBusyLockAsync(busyLock, () -> {
+            if (!event.single()) {
+                // We don't need to handle keys on index creation or deletion.
+                return completedFuture(null);
+            }
+
+            Entry entry = event.entryEvent().newEntry();
+
+            if (!entry.tombstone()) {
+                // In case an index was created when there was only one partition.
+                return completedFuture(null);
+            }
+
+            String partitionBuildIndexKey = new String(entry.key(), UTF_8);
+
+            int indexId = parseIndexIdFromPartitionBuildIndexKey(partitionBuildIndexKey);
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
+
+            long metastoreRevision = entry.revision();
+
+            if (isRemainingPartitionBuildIndexKeys(indexId, metastoreRevision)
+                    || isMetastoreKeyAbsent(startBuildIndexKey, metastoreRevision)) {
+                return completedFuture(null);
+            }
+
+            // We can use the latest version of the catalog since we are on the metastore thread.
+            CatalogIndexDescriptor indexDescriptor = getIndexDescriptorStrict(indexId, catalogManager.latestCatalogVersion());
+
+            // We will not wait for the command to be executed, since we will then find ourselves in a dead lock since we will not be able
+            // to free the metastore thread.
+            catalogManager.execute(buildMakeIndexAvailableCommand(indexDescriptor));
```

Comment:
What if the catalog operation fails (due to some unexpected runtime exception in the MetaStorage), and this happens for all nodes processing this event, but this happens inside the future that we don't wait. Won't we lose the update?

Reply:
Also, if it fails inside a future, we'll not see an exception. `whenComplete()` might be useful to log it.

Reply:
> What if the catalog operation fails (due to some unexpected runtime exception in the MetaStorage), and this happens for all nodes processing this event, but this happens inside the future that we don't wait. Won't we lose the update?

I think that if the operation cannot be performed for reasons other than: lack of index, already available, or the node is stopped, then this is a problem and the error handler will have to fail the node from the cluster.

> Also, if it fails inside a future, we'll not see an exception. whenComplete() might be useful to log it.

I'll add logging.

Reply:
Then I suggest adding a TODO about Failure Handler here (or will it catch the exception in another place?) as we rely on Failure Handler in many places, it will be really easy to forget some of them :)